src/state_migration/mod.rs

```diff
@@ -144,5 +148,70 @@ where
     Ok(None)
 }
 
+pub struct BlockstoreWithWriteBuffer<DB: Blockstore> {
+    inner: DB,
+    buffer: RwLock<HashMap<Cid, Vec<u8>>>,
+    buffer_capacity: usize,
+}
+
+impl<DB: Blockstore> Blockstore for BlockstoreWithWriteBuffer<DB> {
+    fn get(&self, k: &Cid) -> anyhow::Result<Option<Vec<u8>>> {
+        if let Some(v) = self.buffer.read().get(k) {
+            return Ok(Some(v.clone()));
+        }
+        self.inner.get(k)
+    }
+
+    fn has(&self, k: &Cid) -> anyhow::Result<bool> {
+        Ok(self.buffer.read().contains_key(k) || self.inner.has(k)?)
+    }
+
+    fn put_keyed(&self, k: &Cid, block: &[u8]) -> anyhow::Result<()> {
+        {
+            let mut buffer = self.buffer.write();
+            buffer.insert(*k, block.to_vec());
+        }
+        self.flush_buffer_if_needed()
+    }
+}
+
+impl<DB: Blockstore> BlockstoreWithWriteBuffer<DB> {
+    pub fn new(inner: DB) -> Self {
+        Self::new_with_capacity(inner, 10000)
```

Comment:
Let's have it configurable so that operators with more memory in their box can potentially increase this. It'd be great to note how much memory increase should one expect with every 10k buffer increment.